### DISCUSSION TITLE
Problem with CodeQL and the Static Tools Logo Test

### DISCUSSION THREAD

**Discussion by thieshdm** [github/codeql#8356] [Q&A] [answered]
I have written a kernel mode driver for OS Microsoft Windows 10 and now I want to certify it. One Part of the Windows Hardware Lab Kit certification is generating the Driver Verification Log using Visual Studio 2019. The Driver Verification Log consists of 3 parts: Code Analysis Log, Static Driver Verifier Log and CodeQL/Semmle Sarif Log. My problem here is the generation of the CodeQL/Semmle Sarif Log. Microsoft has a description of this on the page: https://docs.microsoft.com/en-us/windows-hardware/drivers/devtest/static-tools-and-codeql
The example from Microsoft and my own driver always break off with the same error:
[ERROR] database finalize> No source code was seen during the build.
                              This can occur if the specified build commands failed to compile or process any code.
                               - Confirm that there is some source code for the specified language in the project.
                               - For codebases written in Go, JavaScript, TypeScript, and Python, do not specify 
                                 an explicit --command.
                               - For other languages, the --command must specify a "clean" build which compiles 
                                 all the source code files without reusing existing build artefacts.
Error code in the log file databases\swayeth1\logbuild-tracer.log :
[E 11:50:35 428] Command: C:\codeql-home\codeql\cpp\tools\win64/extractor.exe --mimic "C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.29.30133\bin\Hostx86\x64\cl.exe" @C:\Users\SCHWEGER\AppData\Local\Temp\tmp2c8c7b4abb5a4ba3a8d93098155c8001.rsp
Warning: C:\Program Files (x86)\Microsoft Visual Studio\2019\Professional\VC\Tools\MSVC\14.29.30133\bin\Hostx86\x64\cl.exe does not look like a compiler. If it is infact a compiler, please contact GitHub.
[E 11:50:55 428] Not a compiler; exiting. Reason: --version stdout was:
--version stderr was:
Microsoft (R) C/C++-Optimierungscompiler Version 19.29.30140 fr x64
Copyright (C) Microsoft Corporation. Alle Rechte vorbehalten.
cl : Befehlszeile warning D9002 : Unbekannte Option "--version" wird ignoriert.
cl : Befehlszeile error D8003 : Name der Quelldatei fehlt.
process exited with code 2

This happens with the CodeQL-version 2.4.6 prescribed by Microsoft. As a test, I tried the CodeQL 2.8.2 version, with this there is no termination. However, that doesn't help me because the CodeQL 2.8.2 version is incompatible with Visual Studio 2019. According to my tests, the program "extractor.exe" is the cause of the problem (directory: C:\codeql-home\codeql\cpp\tools\win64). So I need a problem solution for CodeQL version 2.46.

My development environment consists of the following components:
Microsoft Visual Studio 2019 Version 16.11.10
SDK 10.0.22000.0
WDK 10.0.21381.1

I hope this is the right place for my problem, thank you.

**Comment by smowton** [comment]
~This is happening because we're looking for the whole word `C/C++` in the `--version` output, but in German it hyphenates the word `C/C++-Optimierungscompiler`. If you're able to get it to output in English, likely by setting the system locale of the machine doing the building, that should work around the issue.~

**Reply by smowton** [reply]
Sorry, ignore me, we consider non-alphabetic characters such as - a word delimiter.

**Comment by alexet** [comment]
CodeQL version 2.4.6 has a bug relating to detecting compilers in non-english version of windows which was fixed in codeql 2.8.0 . Changing the language is a workaround.

Can you explain in more detail why you think CodeQL 2.8.2 is incompatible with Visual Studio 2019?

**Comment by thieshdm** [comment]
Thanks for the reply, I will test the workaround with the Windows language switch to English.

Statement regarding the incompatibility of CodeQL 2.8.2 with Visual Studio and Windows Hardware Labor Kit from my point of view: In both cases, the file format is labeled as incorrect. Furthermore, the Microsoft documentation explicitly refers to version 2.4.6 of CodeQL. If the version didn't matter then Microsoft would probably recommend the latest version, but that's not the case.

**Comment by thieshdm** [comment, marked as answer]
Switching the operating system language from German to English works. The test with it was successful. Thanks for the help here, it solved my problem.
